pwndbg/commands/msr.py

```diff
@@ -0,0 +1,146 @@
+parser.add_argument(
+    "-n",
+    "--name",
+    metavar="name",
+    help="MSR litteral value (eg. MSR_STAR)",
+    type=str,
+    nargs="?",
+    default=None,
+)
+parser.add_argument(
+    "-w",
+    "--write",
+    metavar="write_value",
+    help="value to write in MSR",
+    type=int,
+    nargs="?",
+    default=None,
+)
+parser.add_argument(
+    "-l",
+    "--list-msr",
+    action="store_true",
+    help="list common MSRs for the current arch",
+)
+
+# Taken from linux/arch/x86/include/asm/msr-index.h
+# Arbitrary list, feel free to add more.
+X86_MSRS = {
+    "MSR_EFER": 0xC0000080,
+    "MSR_STAR": 0xC0000081,
+    "MSR_LSTAR": 0xC0000082,
+    "MSR_CSTAR": 0xC0000083,
+    "MSR_SYSCALL_MASK": 0xC0000084,
+    "MSR_FS_BASE": 0xC0000100,
+    "MSR_GS_BASE": 0xC0000101,
+    "MSR_KERNEL_GS_BASE": 0xC0000102,
+    "MSR_TSC_AUX": 0xC0000103,
+}
+
+
+COMMON_MSRS = {"x86": X86_MSRS, "x86-64": X86_MSRS}
+
+
+def x86_msr_read(msr: int) -> None:
+    arch = pwndbg.aglib.arch.name
+
+    if arch == "x86-64":
+        setattr(pwndbg.aglib.regs, "rcx", msr)
+    else:
+        setattr(pwndbg.aglib.regs, "ecx", msr)
+
+    async def ctrl(ec: pwndbg.dbg_mod.ExecutionController):
+        sc = pwnlib.asm.asm("rdmsr")
+        async with pwndbg.aglib.shellcode.exec_shellcode(ec, sc):
+            edx = int(pwndbg.aglib.regs["edx"]) << 32
+            eax = int(pwndbg.aglib.regs["eax"])
+            ret = edx + eax
+            print(hex(ret))
+
+    pwndbg.dbg.selected_inferior().dispatch_execution_controller(ctrl)
+
+
+def x86_msr_write(msr: int, write_value: int) -> None:
+    arch = pwndbg.aglib.arch.name
+
+    eax = write_value & 0xFFFFFFFF
+    edx = write_value >> 32
+    if arch == "x86-64":
+        setattr(pwndbg.aglib.regs, "rcx", msr)
+        setattr(pwndbg.aglib.regs, "rax", eax)
+        setattr(pwndbg.aglib.regs, "rdx", edx)
+    else:
+        setattr(pwndbg.aglib.regs, "ecx", msr)
+        setattr(pwndbg.aglib.regs, "eax", eax)
+        setattr(pwndbg.aglib.regs, "edx", edx)
+
+    async def ctrl(ec: pwndbg.dbg_mod.ExecutionController):
+        sc = pwnlib.asm.asm("wrmsr")
+        async with pwndbg.aglib.shellcode.exec_shellcode(ec, sc):
+            return
+
+    pwndbg.dbg.selected_inferior().dispatch_execution_controller(ctrl)
+
+
+def msr_read(msr: int) -> None:
+    arch = pwndbg.aglib.arch.name
+
+    if arch == "x86" or arch == "x86-64":
+        x86_msr_read(msr)
+    else:
+        print(f"{arch} not supported")
+
+
+def msr_write(msr: int, write_value: int) -> None:
+    arch = pwndbg.aglib.arch.name
+
+    if arch == "x86" or arch == "x86-64":
+        x86_msr_write(msr, write_value)
+    else:
+        print(f"{arch} not supported")
+
+
+def msr_list(arch: str):
+    msrs = COMMON_MSRS[arch]
+    for msr, value in msrs.items():
+        print(f"{msr}\t{hex(value)}")
+
+
+@pwndbg.commands.ArgparsedCommand(parser, category=CommandCategory.KERNEL, command_name="msr")
+@pwndbg.commands.OnlyWhenQemuKernel
```

Comment:
Nitpick: what if I do bare metal debugging of linux kernel through e.g. [kgdboe](https://github.com/sysprogs/kgdboe)?

Then I may also want to use this command. Maybe we just need some warning and yes/no question within the command if its not qemu kernel to ask the user if they want to proceed?

Also, this can be "OnlyWithArch x86/x86-64" or something for now.

Reply:
most of commands we have guarded with `OnlyWhenQemuKernel` :(

Reply:
Lets leave OnlyWhenQemuKernel here for now.